### PR TITLE
feat(ai-provider): watcher cron scheduling via croner (Spec 45)

### PR DESCRIPTION
**Pull request by laofahai**
## What

Gives watchers a real **time-based (cron) trigger** (Spec 45). Previously the
watcher engine evaluated on every poll; this adds deterministic cron-driven
scheduling.

- `ScheduleTracker` (croner) parses each watcher's cron expression and computes
  the next fire time in **UTC**.
- The engine learns when a scheduled tick is due via `runScheduleTick()` /
  `getNextScheduledRun()`, with an injectable `clock?: () => Date` so schedule
  logic is fully deterministic under test (no wall-clock dependency).
- Count-condition watchers resolve their threshold via `queryRecords`.

## Files

| File | Change |
|------|--------|
| `src/watcher-schedule.ts` | **new** — `ScheduleTracker`: cron parse + next-run computation |
| `src/watcher-engine.ts` | inject `clock` + `scheduleIntervalMs`; add `runScheduleTick()` / `getNextScheduledRun()` |
| `__tests__/watcher-schedule.test.ts` | **new** — tracker + engine tick tests |
| `package.json` / `bun.lock` | `croner ^10.0.1` (already an approved repo dep — core depends on it) |

## Testing

- `bun run check` (Biome 2.4.16) — clean
- `bun run typecheck` — exit 0
- Targeted `bun test addons/ai-provider/cap-ai-provider/` — **290 pass / 3 skip** (E2E, no API key) **/ 0 fail**
- Full `bun run test` runs in CI as the authoritative backstop.

## Note

`watcher-engine.ts` is now ~864 lines (over the 500-line guideline). The new
schedule logic itself lives in the separate `watcher-schedule.ts`; the engine
growth is integration glue. A follow-up to split `watcher-engine.ts` by
responsibility can be tracked separately.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for schedule-based watchers using cron expressions to trigger actions at specified times.
  * New methods to query and execute scheduled watcher runs.

* **Tests**
  * Added comprehensive test suite for cron schedule trigger behavior, including edge cases and error handling.

### CODE REVIEW

File: addons/ai-provider/cap-ai-provider/src/watcher-engine.ts [hunk 1]
```diff
@@ -296,6 +365,111 @@ class WatcherEngineImpl implements WatcherEngine {
...
+              context: {},
+              actor: { type: "system", id: "watcher", groups: [] },
+            }),
+          )
+        : records;
+
+      const count = matching.length;
+      if (!evaluateComparison(count, countCondition)) {
+        return { watcherName: watcher.name, fired: false, reason: "count_condition_not_met" };
+      }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply addons/ai-provider/cap-ai-provider/src/watcher-engine.ts]: Fixed in 97f9a63 — when the count condition is not met, the schedule evaluator now resets the debounce state to false before returning count_condition_not_met, so an once_until_reset watcher fires again when the condition later becomes true. Test added (fires → debounced → count drops/resets → fires again).

File: addons/ai-provider/cap-ai-provider/src/watcher-engine.ts [hunk 2]
```diff
@@ -236,6 +275,36 @@ class WatcherEngineImpl implements WatcherEngine {
...
+    const results: WatcherEvaluationResult[] = [];
+
+    for (const watcherName of dueWatcherNames) {
+      // Re-check enabled status on every occurrence — a watcher may have been
+      // disabled or removed since the schedule was registered.
+      const watcher = this.registry.get(watcherName);
+      if (!watcher?.enabled || watcher.trigger.type !== "schedule") {
+        results.push({ watcherName, fired: false, reason: "watcher_unavailable" });
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply addons/ai-provider/cap-ai-provider/src/watcher-engine.ts]: Fixed in 97f9a63 — a watcher detected unavailable/disabled during a tick is now removed from the ScheduleTracker (tracker.remove), stopping endless next-run computation. Test asserts getNextScheduledRun returns undefined afterwards.

File: addons/ai-provider/cap-ai-provider/src/watcher-engine.ts [hunk 3]
```diff
@@ -296,6 +365,111 @@ class WatcherEngineImpl implements WatcherEngine {
...
+    const groupKey = "schedule";
+    if (!this.shouldFire(watcher, groupKey, true)) {
+      return { watcherName: watcher.name, fired: false, reason: "debounced" };
+    }
+
+    await this.fireEffect(watcher, ctx);
+    this.updateState(watcher.name, groupKey, true);
+
+    return { watcherName: watcher.name, fired: true };
+  }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply addons/ai-provider/cap-ai-provider/src/watcher-engine.ts]: Fixed in 97f9a63 — watcher-engine.ts went 865 → 446 lines by extracting schedule-engine.ts (253) and mutation-engine.ts (375); collaborators passed via deps objects, all public WatcherEngine signatures unchanged (they delegate). All files now under the 500-line limit.

File: addons/ai-provider/cap-ai-provider/src/watcher-engine.ts [hunk 4]
```diff
@@ -296,6 +365,111 @@ class WatcherEngineImpl implements WatcherEngine {
...
+        );
+      }
+    }
+
+    // All cron expressions failed to register — nothing to poll.
+    if (registered === 0) return;
+
+    const interval = setInterval(() => {
+      void this.runScheduleTick();
+    }, this.scheduleIntervalMs);
```
[the remark this answers is not in the document]
Comment (laofahai) [reply addons/ai-provider/cap-ai-provider/src/watcher-engine.ts]: Fixed in 97f9a63 — runTick() now guards on an in-flight promise; a concurrent call returns the identical in-flight pass and self-clears on settle, so a slow fireEffect() can no longer let setInterval start a second pass that re-fires the same due occurrence. Test added (overlapping ticks, slow effect → effect runs once).

File: addons/ai-provider/cap-ai-provider/src/watcher-schedule.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        state.nextDue.getTime() <= now.getTime() &&
+        guard < maxOccurrencesPerWatcher
+      ) {
+        due.push(state.watcherName);
+        // Advance strictly past the consumed occurrence so the same tick is
+        // never reported twice. croner's nextRun is exclusive of `prev`.
+        state.nextDue = state.cron.nextRun(state.nextDue);
+        guard += 1;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (laofahai) [reply addons/ai-provider/cap-ai-provider/src/watcher-schedule.ts:119]: Fixed in 97f9a63 — collectDue() now accumulates { watcherName, dueAt }, sorts ascending by dueAt, and maps back to names, so occurrences come back chronologically across watchers (not Map/registration order). Test added (later-due watcher registered first → results sorted by due time).
